File: app/interfaces/json/project/ProjectResponse.scala
package interfaces.json.project

import dev.tchiba.sdmt.core.models.project.Project
import io.circe.generic.semiauto.deriveEncoder
import io.circe.syntax._
import io.circe.{Encoder, Json}

final class ProjectResponse private (project: Project) {
  private case class Response(
      projectId: String,
      projectAlias: String,
      projectName: String,
      projectOverview: String
  )
  implicit private val encoder: Encoder[Response] = deriveEncoder[Response]
  private val response = Response(
    projectId = project.id.asString,
    projectAlias = project.alias.value,
    projectName = project.name,
    projectOverview = project.overview
  )

  def json: Json = response.asJson
}

object ProjectResponse {
  def apply(project: Project) = new ProjectResponse(project)
}

File: front/angular-front/src/app/models/project/http/ProjectResponse.ts
import {Project} from "../project";
import {ProjectId} from "../id/project-id";
import {ProjectAlias} from "../alias/project-alias";
import {ProjectName} from "../name/project-name";
import {ProjectOverview} from "../overview/project-overview";


export type ProjectResponse = {
  readonly projectId: string
  readonly projectAlias: string
  readonly projectName: string
  readonly projectOverview: string
}

export const ProjectResponse = {
  convert(res: ProjectResponse): Project {
    return new Project(
      new ProjectId(res.projectId),
      new ProjectAlias(res.projectAlias),
      new ProjectName(res.projectName),
      new ProjectOverview(res.projectOverview)
    )
  }
}

